### DISCUSSION TITLE
[Feature Request] Generate asset transforms via URL query params

### DISCUSSION THREAD

**Discussion by mylesthedev** [craftcms/cms#12343] [Ideas] [answered]
This feature is more targeted towards those who are using GQL as a way to get content out of Craft.

Similar to how other image services work e.g. S3 / Imgix etc - having the ability to just add a few query params to the end of an image url would mean we can generate images on the fly.

For example `myAmazingImage.jpg?w=1900&h=500&q=95` a request to that url would generate a specific transform.

Having this functionality baked in would strengthen Craft's ability for headless sites.

Until then, does anyone know of a good way to implement this?

**Comment by brianjhanson** [comment, marked as answer]
You can already generate images on the fly via graphql by using the [transform directive](https://craftcms.com/docs/4.x/graphql.html#the-transform-directive) to generate a specific transform. 

For example:
```gql
{
  entries (section: "news", limit: 2, orderBy: "dateCreated DESC") {
    ... on news_article_Entry {
      featuredImage {
        url @transform (width: 300) // operates similarly to the `url()` transform syntax
      }
    }
  }
}
```

For something really URL based, you'd have to make something yourself at the moment. A super simple example could be just to add a custom controller and use that to generate the transform. 

```php
<?php
/**
 * @link https://craftcms.com/
 * @copyright Copyright (c) Pixel & Tonic, Inc.
 * @license https://craftcms.github.io/license/
 */

namespace modules\controllers;

use craft\elements\Asset;
use craft\web\Controller;
use yii\web\NotFoundHttpException;

class TransformController extends Controller
{
    public array|bool|int $allowAnonymous = ['index'];

    public function actionIndex(string $filename)
    {
        $width = $this->request->getParam('w');
        $height = $this->request->getParam('h');

        $asset = Asset::find()->filename($filename)->one();
        if (!$asset) {
            throw new NotFoundHttpException('Asset not found.');
        }

        $url = $asset->getUrl([
            'width' => $width, 
            'height' => $height
        ], true);

        return $this->redirect($url);
    }
}
```

And then in your main module/plugin class, make sure to register the route.

```php
Event::on(
    UrlManager::class,
    UrlManager::EVENT_REGISTER_SITE_URL_RULES,
    function(RegisterUrlRulesEvent $event) {
        $event->rules['transform/<filename:.+>'] = 'my-module/transform/index';
    }
);
````

You'd probably want to implement some kind of token system as well to make sure you're the only one requesting transforms.
